### PR TITLE
Handle shared modules in discovery better

### PR DESCRIPTION
**Pull request by michalsidzej**
Resolves L2B-1331
This PR will make it clear for the users that a change in a shared module (rn only l2beat-starkware) is effectively a change in all of the dependant projects. The tests make sure, that contracts marked as shared are only discovered once. It also automatically detects all the dependants of a module, so the message is clear and doesn't have to be hardcoded.

Screenshot 2023-04-14 at 09 14 33

### CODE REVIEW

File: packages/backend/src/core/discovery/utils/diffToMessages.ts [hunk 1]
```diff
@@ -13,7 +19,9 @@ export function diffToMessages(name: string, diffs: DiscoveryDiff[]): string[] {
 
   const bundledMessages = bundleMessages(messages, maxLength)
 
-  return bundledMessages.map((m) => `${header}${wrapDiffCodeBlock(m)}`)
+  return bundledMessages.map(
+    (m) => `${header}${dependentsMessage}${wrapDiffCodeBlock(m)}`,
```
Comment (antooni) [inline packages/backend/src/core/discovery/utils/diffToMessages.ts:23]: Have you tried how this message will look like in Discord? Please provide a screenshot in PR description

File: packages/backend/src/core/discovery/utils/config.test.ts [hunk 2]
```diff
@@ -128,7 +128,23 @@ describe('discovery config.jsonc', () => {
           }
         }
       }
-    })
+    }),
+      it('all shared modules exist', async () => {
```
Comment (antooni) [inline packages/backend/src/core/discovery/utils/config.test.ts:132]: good one

File: packages/backend/discovery/apex/config.jsonc [hunk 3]
```diff
@@ -50,7 +50,7 @@
       }
     },
     "SHARPVerifierProxy": {
-      "ignoreDiscovery": true
+      "sharedModule": "l2beat-starkware"
```
Comment (antooni) [inline packages/backend/discovery/apex/config.jsonc:53]: or maybe add a field "sharedModules" to the config, which will be the list of the contract identifiers?

Comment (michalsidzej) [reply packages/backend/discovery/apex/config.jsonc:53]: Do you mean config.ts?

File: packages/backend/src/core/discovery/utils/diffToMessages.ts (hunk 1 above)
Comment (michalsidzej) [reply packages/backend/src/core/discovery/utils/diffToMessages.ts:23]: Updated

File: packages/backend/discovery/apex/config.jsonc [hunk 4]
```diff
@@ -8,6 +8,9 @@
     "0x23Cab3CF1aa7B929Df5e9f3712aCA3A6Fb9494E4": "Committee",
     "0x47312450B3Ac8b5b8e247a6bB6d523e7605bDb60": "SHARPVerifierProxy"
   },
+  "sharedModules": {
```
Comment (antooni) [inline packages/backend/discovery/apex/config.jsonc:11]: <3

File: packages/backend/src/core/discovery/config/DiscoveryConfig.ts [hunk 5]
```diff
@@ -32,7 +32,25 @@ export class DiscoveryConfig {
...
+    return this.config.sharedModules
+      ? Object.values(this.config.sharedModules)
+      : []
+  }
+
   get hash(): Hash256 {
     return hashJson(getDiscoveryConfigEntries(this.config))
   }
+
+  getSharedModule(address: EthereumAddress): string | undefined {
```
Comment (antooni) [inline packages/backend/src/core/discovery/config/DiscoveryConfig.ts:45]: maybe isSharedModule?

File: packages/backend/src/core/discovery/runDiscovery.ts [hunk 6]
```diff
@@ -70,7 +71,11 @@ export async function dryRunDiscovery(
   )
 
   if (diff.length > 0) {
-    const messages = diffToMessages(projectConfig.name, diff)
+    const messages = diffToMessages(
```
Comment (antooni) [inline packages/backend/src/core/discovery/runDiscovery.ts:74]: Some performance idea, run findDependants only if the name starts with 'l2beat-'

**Review by antooni — CHANGES_REQUESTED** [review]
please address my comments

File: packages/backend/src/core/discovery/config/DiscoveryConfig.ts (hunk 5 above)
Comment (michalsidzej) [reply packages/backend/src/core/discovery/config/DiscoveryConfig.ts:45]: I mean this gets `SharedModule | undefined', decided not to create two functions like this.

File: packages/config/src/discovery/ProjectDiscovery.test.ts [hunk 7]
```diff
@@ -89,6 +89,18 @@ describe(ProjectDiscovery.name, () => {
           `Assertion Error: Upgradeability param of key ${key} does not exist in ${contractStub.name} contract (${projectName})`,
         )
       })
+
+      it('should not throw an error if upgradeability param is 0', () => {
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/config/src/discovery/ProjectDiscovery.test.ts:93]: [Ex. when you have upgradeDelay: 0 ](https://github.com/l2beat/l2beat/blob/master/packages/backend/discovery/dydx/discovered.json#L13)

File: packages/config/src/discovery/ProjectDiscovery.ts [hunk 8]
```diff
@@ -144,7 +144,7 @@ export class ProjectDiscovery {
     const contract = this.getContract(contractIdentifier)
     const result = contract.values?.[key] as T | undefined
     assert(
-      result,
+      isNonNullable(result),
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/config/src/discovery/ProjectDiscovery.ts:147]: Nothing really, I just spotted a small bug, when getting an upgradeability param that's 0 we get an assertion error. This is why i wrote a test for that

**Review by antooni — APPROVED** [review]

File: packages/backend/src/core/discovery/runDiscovery.ts (hunk 6 above)
Comment (michalsidzej) [reply packages/backend/src/core/discovery/runDiscovery.ts:74]: This is happening already, first line inside the `findDependants` function

**Pull request merged**